File: core/smap.fs
module core.smap

open System
open System.Collections
open System.Collections.Generic

type SMap<'T, 'U when 'T : comparison> =
    {
        data : Map<'T, 'U>
    }

module SMap =
    let ofList<'a, 'b when 'a : comparison> (as' : ('a * 'b) list) : SMap<'a, 'b> =
        {
            data = 
                as'
                |> Map.ofList
        }

    let ofLookup<'a, 'b when 'a : comparison> (as' : 'a list) (fn : 'a -> 'b) : SMap<'a, 'b> =
        {
            data = 
                as'
                |> List.map (fun a -> (a, fn a))
                |> Map.ofSeq
        }

    let get<'a, 'b when 'a : comparison> (sm : SMap<'a, 'b>) (a : 'a) : 'b =
        sm.data.Item a

    let tryGet<'a, 'b when 'a : comparison> (sm : SMap<'a, 'b>) (a : 'a) : 'b option =
        sm.data.TryFind a
(*
    let get<'a, 'b when 'a : comparison> (as' : ('a * 'b) list) (k : 'a) : 'b =
        as'
        |> List.find (fst >> ((=) k))
        |> snd
*)

File: src/smap.ml
let ofLookup (fn : 'a -> 'b) (as' : 'a list) : ('a * 'b) list =
    List.map (fun a -> (a, fn a)) as'

let get (comparer : 'a -> 'a -> int) (k : 'a) (as' : ('a * 'b) list) : 'b =
    as'
    |> List.find (fun a -> comparer k (fst a) = 0)
    |> snd

let tryGet (comparer : 'a -> 'a -> int) (k : 'a) (sm : ('a * 'b) list) : 'b option =
    if List.exists (fun a -> comparer k (fst a) = 0) sm then Some (get comparer k sm)
    else None
